### ISSUE TITLE
Proposal: estree/tests

### ISSUE THREAD

**Issue by RReverser** [estree/estree#55]
Now that we have most of ES6-related issues resolved, it became obvious that every implementor benefits from working together on a single spec, as developers can focus on making tools better, cleaner and faster instead of resolving interop issues.

I think that next step towards making interop easier would be creating repo under ESTree org with automated tests with exported API that could be used by any spec-compliant parser. That way, any representation that was already agreed in this repo, will be reflected in tests, and parsers could use them either as `devDependency` or `git submodule` for making sure that everyone parses things in the same way or throws error when needed (that said, tests should be still flexible enough to allow different implementors to throw different error messages and not to include parser-specific node extensions).

Thoughts?

/cc @mikesherov @ariya @dherman

**Comment by mikesherov** [comment]
Does that mean we create an IDL?

**Comment by michaelficarra** [comment]
I think the more useful project would be a schema validator, so no matter how you generate an AST, you can make sure it conforms to the spec. In order to use this for testing a parser, you could run it against the output of each of the parser's tests or just the result of parsing [everything.js](https://github.com/michaelficarra/everything.js) (ES6 support should be complete later today).

**Comment by mikesherov** [comment]
@michaelficarra that sounds great to me.

**Comment by RReverser** [comment]
Yes, something like `everything.js`-based test would be a good start.

In wider meaning, I rather thought about complete test suite - initially Acorn tests anyway derived from Esprima tests, and now both evolved with new edge cases / ESTree changes and so on, and I thought that both teams (and probably even `Reflect.parse`) could benefit from merging and sharing same test suite via npm package or git submodule with exposed DI interface for `test` callbacks.

**Comment by mikesherov** [comment]
Oh, like http://test262.ecmascript.org/ I think that would be very useful.

**Comment by RReverser** [comment]
> Oh, like http://test262.ecmascript.org/

Exactly! :)

**Comment by nzakas** [comment]
:+1:

**Comment by impinball** [comment]
If I find time, I was thinking of creating a custom WebIDL-based validation library, for this very thing. This has also come up in #60.

**Comment by michaelficarra** [comment]
@RReverser I spoke to @bterlson last night about his interests, and he said he would be willing to accept general parsing pass/fail tests into test262. We should put together a PR to add the necessary framework portions needed to allow parser authors to pull their tests into test262. We would all benefit from this. /cc @ariya @ikarienator

**Comment by bterlson** [comment]
Before putting together a PR, instead it would probably be better to write a proposal first. Good to tease out issues before devoting much time :)

**Comment by michaelficarra** [comment]
Sure. This could be a good topic for the upcoming jQuerySF face-to-face meeting.
